### ISSUE TITLE
Changing --action_env doesn't cause actions to be rerun

### ISSUE THREAD

**Issue by kchodorow** [bazelbuild/bazel#3265]
E.g.,

```
$ bazel test --test_output=all --action_env=TZ=Local :tz_test 
...
timezone: Local
================================================================================
Target //:tz_test up-to-date:
  bazel-bin/tz_test
INFO: Elapsed time: 1.034s, Critical Path: 0.05s
//:tz_test                                                               PASSED in 0.0s

Executed 1 out of 1 test: 1 test passes.
$ bazel test --test_output=all --action_env=TZ=Whatever :tz_test 
...
timezone: Local
================================================================================
//:tz_test                                                      (cached) PASSED in 0.0s
```

The only way to get it to pick up on the new value is by doing a clean in-between.

**Comment by aehlig** [comment]
> The only way to get it to pick up on the new value is by doing a clean in-between.

And I believe even then it wouldn't pick up the new value. For some strange historic
reasons (with some strong opinions upon) the test-environment does not inherit from
the action environment. Does using ``--test_env`` have the desired effect?

-- 
Klaus Aehlig
Google Germany GmbH, Erika-Mann-Str. 33, 80636 Muenchen
Registergericht und -nummer: Hamburg, HRB 86891
Sitz der Gesellschaft: Hamburg
Geschaeftsfuehrer: Matthew Scott Sucherman, Paul Terence Manicle

**Comment by aehlig** [comment]
On Mon, Jun 26, 2017 at 05:09:43PM +0200, Klaus Aehlig wrote:
> > The only way to get it to pick up on the new value is by doing a clean in-between.
> 
> And I believe even then it wouldn't pick up the new value. For some strange historic
> reasons (with some strong opinions upon) the test-environment does not inherit from
> the action environment. Does using ``--test_env`` have the desired effect?

Oh, sorry. It seems that it does get picked up, just not by the cache key. I'll try to
find a fix for this. The ``--test_env`` is however honored by tests; unfortunately, the
``action_env`` influences the tests as well, without being noticed.

-- 
Klaus Aehlig
Google Germany GmbH, Erika-Mann-Str. 33, 80636 Muenchen
Registergericht und -nummer: Hamburg, HRB 86891
Sitz der Gesellschaft: Hamburg
Geschaeftsfuehrer: Matthew Scott Sucherman, Paul Terence Manicle

**Comment by aehlig** [comment]
> Oh, sorry. It seems that it does get picked up, just not by the cache key. I'll try to
> find a fix for this. The ``--test_env`` is however honored by tests; unfortunately, the
> ``action_env`` influences the tests as well, without being noticed.

Interestingly enough, with ``--action_env=TZ`` the dependency on $TZ is correctly tracked.
So the wrong tracking is in the client-environment-independent ("static") part of the
test-action-environment.


-- 
Klaus Aehlig
Google Germany GmbH, Erika-Mann-Str. 33, 80636 Muenchen
Registergericht und -nummer: Hamburg, HRB 86891
Sitz der Gesellschaft: Hamburg
Geschaeftsfuehrer: Matthew Scott Sucherman, Paul Terence Manicle

**Comment by aehlig** [comment]
Interestingly enough, that issue is already fixed by d1c5329ba622b29afd3ab9f670fa17064d493bc0. However, it went unnoticed for a long time. Adding a test to avoid that breakage in the future, see https://bazel-review.googlesource.com/c/12491/
